src/Host/Client/Impl/Session/RSessionProvider.cs

```diff
@@ -172,31 +187,22 @@ public class RSessionProvider : IRSessionProvider {
                 lockToken = await _brokerSwitchLock.WaitAsync(cancellationToken);
                 await _connectCde.WaitAsync(cancellationToken);
             } catch (OperationCanceledException) {
+                brokerClient.Dispose();
                 return false;
             }
 
+            // First switch broker proxy so that all new sessions are created for the new broker
+            var oldBroker = _brokerProxy.Set(brokerClient);
             try {
-                // First switch connector so that all new sessions are created for the new broker
-                var oldBroker = _brokerProxy.Set(brokerClient);
                 BrokerChanging?.Invoke(this, EventArgs.Empty);
-                var switchingFromNull = oldBroker is NullBrokerClient;
-                if (!switchingFromNull) {
-                    _callback.WriteConsole(Resources.RSessionProvider_StartSwitchingWorkspaceFormat.FormatInvariant(_brokerProxy.Name, GetUriString(_brokerProxy)));
-                }
-
-                var sessions = _sessions.Values.ToList();
-                if (sessions.Any()) {
-                    var sessionsSwitched = await TrySwitchSessionsAsync(sessions, oldBroker, cancellationToken);
-                    if (!sessionsSwitched) {
-                        BrokerChangeFailed?.Invoke(this, EventArgs.Empty);
-                        return false;
-                    }
-                }
-
-                if (!switchingFromNull) {
-                    _callback.WriteConsole(Resources.RSessionProvider_SwitchingRWorkspaceCompleted);
-                }
+                await SwitchBrokerAsync(cancellationToken, oldBroker);
+                oldBroker.Dispose();
                 PrintBrokerInformation();
+            } catch(Exception) {
```

Comment:
Yeah, probably I should should rethrow exceptions that aren't OCE

---
In reply to: [80101432](https://github.com/Microsoft/RTVS/pull/2384#discussion_r80101432) [](ancestors = 80101432)